README.md
# @kxgo/easy-mock

`easy-mock`只需一些简单的配置就可以轻松`mock`数据

### 安装
--------
```
npm i @kxgo/easy-mock
or
yarn add @kxgo/easy-mock
```

### 使用
--------
``` javascript
const path = require('path')
const easyMock = require('@kxgo/easy-mock')

new easyMock({
  // 请传递完整的配置文件路径进去，服务是会自动刷新
  path: path.resolve(__dirname, 'xxxx.js')
})
```
`配置文件的例子`
``` javascript
// xxxx.js
module.exports = {
  // common选项不是必须, 内置的配置如下，当然你也可以更改
  common: {
    // mock服务的默认端口，如果端口被占用，会换一个
    port: 8018,
    // 如果你想看一下ajax的loading效果，该配置项可以设置接口的返回延迟
    timeout: 500,
    // 如果你想看一下接口请求失败的效果，将rate设置成0就可以了，rate取值范围0~1，代表成功的概率
    rate: 1,
    // mock属性为false的话可以关闭掉mock服务
    mock: true,
    // 是否打印调试信息
    debug: false
  },
  // api...，可以不用写全，写最后一个就好
  'getList': {
    code: 0,
    'data|5': [{
      'uid|1000-99999': 999,
      'name': '@cname'
    }],
    result: true
  },
  '/pkApi/getPerson': {
    code: 0,
    'data|5': [{
      'name': '@cname'
    }],
    result: true
  },
  '/pkApi/getOther': {
    code: 0,
    'data|5': [{
      'name': '@cname'
    }],
    result: true
  }
}
```
服务启动后，打开`http://127.0.0.1:8018/pkApi/getPerson`就可以看到mock的数据了！

### 本地debug
```
npm i kxg -g
npm i nodemon -g

npm run dev
```

### 发布
```
npm i standard-version -g
npm run pub
```

### Q&A
--------
##### 1 怎么去mock jsonp
在链接后面传`callback`参数就好了
##### 2 怎么去mock数据
参考[mock rule](http://mockjs.com/examples.html)
##### 3 怎么去处理更复杂的场景，我想要自定义数据返回
``` javascript
module.exports = {
  ...
  ['/pkApi/getOther'] (req, res) {
    const query = req.query
    // 可以自己写处理逻辑
    return {}
  }
  ...
}
```
`get`类型的请求可以通过`req.query`去获取参数

`post`类型的请求可以通过`req.body`去获取参数

### issues
-----------
[issue list](https://github.com/kxgo/easy-mock/issues)

### changelog

[changelog](./CHANGELOG.md)



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.1.4](https://github.com/nwa2018/easy-mock/compare/v0.1.3...v0.1.4) (2019-02-26)


### Bug Fixes

* options请求正常返回 ([3c846d4](https://github.com/nwa2018/easy-mock/commit/3c846d4))



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.1.3](https://github.com/nwa2018/easy-mock/compare/v0.1.2...v0.1.3) (2019-02-26)



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.1.2](https://github.com/nwa2018/easy-mock/compare/v0.1.1...v0.1.2) (2019-01-14)


### Bug Fixes

* fix common params ([a922955](https://github.com/nwa2018/easy-mock/commit/a922955))



# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 0.1.1 (2019-01-14)


src/index.js
const path = require('path')
const childProcess = require('child_process')
const chokidar = require('chokidar')
const portfinder = require('portfinder')
const EventEmitter = require('events').EventEmitter

module.exports = class EasyMock extends EventEmitter {
  constructor (opts) {
    super()
    if (!opts.path) throw new Error('请设置common.path指定mock配置文件的路径')

    this.path = opts.path
    this.debug = opts.debug
    const options = require(this.path)
    this.init()
  }

  init () {
    const options = require(this.path)
    if (options.common) {
      if (options.common.mock !== undefined && !options.common.mock) {
        this.debug && console.log('options mock is false, stop the mock server...')
        return
      }
    }
    this.port = (options.common && options.common.port) || 8018
    portfinder.basePort = this.port
    portfinder.getPort((err, port) => {
      if (err) {
        throw new Error(err)
      }
      this.port = port
      this.setUpWatcher()
    })
  }

  setUpWatcher () {
    this.forkChild()
    chokidar.watch(this.path, {
      persistent: true
    }).on('change', _ => {
      this.emit('killChild')
      setTimeout(_ => {
        this.forkChild()
      }, 30)
    })
  }

  forkChild () {
    const child = childProcess.fork(path.join(__dirname, 'server.js'), [], {
      encoding: 'utf8',
      execArgv: process.execArgv
    })
    child.on('error', (data) => {
      this.debug && console.log(`Express server exited with error ${data.toString()}`)
    })

    child.on('exit', (code) => {
      this.debug && console.log(`Express server exited with code ${code || 0}`)
    })

    this.on('killChild', _ => {
      child.kill('SIGKILL')
      this.removeAllListeners()
    })
    // 给子进程传递数据
    child.send({
      path: this.path,
      port: this.port,
      debug: this.debug
    })
    return child
  }
}


src/middlewares.js
exports.successRate = rate => (req, res, next) => {
  if (rate > Math.random()) return next()
  return next(500)
}

exports.delayRes = time => (req, res, next) => {
  setTimeout(function() { next() }, time)
}

exports.jsonRes = () => (req, res, next) => {
  if (req.method === 'get') {
    res.setHeader('Content-Type', 'application/json')
  }
  next()
}

exports.crossDomain = (config) => (req, res, next) => {
  const {
    allowHeaders = '',
    withCredentials = true
  } = config
  const baseHeader = "Origin, X-Requested-With, Content-Type, Accept, Authorization"
  res.header("Access-Control-Allow-Origin", req.headers.origin);
  res.header("Access-Control-Allow-Methods", "GET,HEAD,OPTIONS,POST,PUT");
  if (withCredentials) {
    res.header('Access-Control-Allow-Credentials', 'true')
  }
  res.header("Access-Control-Allow-Headers", allowHeaders ? `${baseHeader}, ${allowHeaders}` : baseHeader);

  if (req.method === 'OPTIONS') res.status(200).json({
    code: 0
  })
  next();
}


src/server.js
const express = require('express')
const bodyParser = require('body-parser')
const { mock } = require('mockjs')

const {
  crossDomain,
  successRate,
  delayRes,
  jsonRes
} = require('./middlewares')

const commonConfig = {
  timeout: 500,
  rate: 1
}

process.on('message', ({path, port, debug}) => {
  // 清除require缓存，每次都可以获取最新数据
  delete require.cache[path]
  const options = require(path)
  // 获取配置数据
  const {
    common = {}
  } = options
  delete options.common

  const {
    rate = 1,
    timeout = 500,
    cross = {}
  } = common

  let dataType

  const app = express()

  // 解析post body数据
  app.use(bodyParser.json())
  app.use(bodyParser.urlencoded({ extended: false }))

  app.use(crossDomain(cross))
  app.use(successRate(rate))
  app.use(delayRes(timeout))
  app.use(jsonRes())

  app.use((req, res, next) => {
    dataType = req.query.callback ? 'jsonp' : 'json'
    next()
  })
  // 路由创建
  Object.keys(options).forEach(path => {
    let data = options[path]
    path = path.indexOf('/') === -1
      ? `*/${path}`
      : `*${path}`
    app.use(path, (req, res, next) => {
      const mockData = typeof data === 'function' ? data(req, res) : data
      res.status(200)[dataType](mock(mockData))
    })
  })

  app.use((req, res, next) => {
    res.status(404).json({ status: 0, code: 404, msg: 'Not Found' })
  })

  app.use((err, req, res, next) => {
    res.status(500).json({ status: 0, code: 500, msg: 'Server Error' })
  })
  app.listen(port, _ => {
    debug && console.log(`Mock Express server listening on port ${port}`)
  })

})
